File: Longest-Palindromic-Substring/Longest-Palindromic-Substring.cpp

class Solution {
public:
    int longestPalindromeHelper(string s,int i,int j)
    {
        int n=s.length();
        while(i>=0&&j<n&&s[i]==s[j])
        {
            i--;
            j++;
        }
        return j-i-1;
    }
    string longestPalindrome(string s) {
        int n=s.length();
        int maxlen=1;
        int startpos=0;
        for(int i=0;i<n;i++)
        {
            int len1=longestPalindromeHelper(s,i,i);
            int len2=0;
            if(i+1<n)
                len2=longestPalindromeHelper(s,i,i+1);
            int curlen=max(len1,len2);
            if(curlen>maxlen)
            {
                maxlen=curlen;
                if(maxlen%2)
                    startpos=i-maxlen/2;
                else
                    startpos=i-(maxlen-1)/2;
            }
        }
        return s.substr(startpos,maxlen);
    }
};


 

File: solutions/Longest-Palindromic-Substring/Longest-Palindromic-Substring.swift
class Solution {
    func longestPalindrome(_ s: String) -> String {
        guard s.count > 1 else {
            return s
        }
        var start = 0
        var end = 0
        let arr = Array(s)
        for (i, c) in arr.enumerated() {
            let ans1 = longestPalindromeHelper(arr, i, i)
            let ans2 = i < arr.count - 1 ? longestPalindromeHelper(arr, i, i + 1) : 0
            let len = max(ans1, ans2)
            if len >= end - start + 1 {
                start = len % 2 == 1 ? i - len / 2 : i - (len - 1) / 2
                end = i + len / 2
            }
        }
        return String(arr[start...end])
    }
    
    private func longestPalindromeHelper(_ s: [Character], _ midLeft: Int, _ midRight: Int) -> Int {
        var i = midLeft
        var j = midRight
        while i >= 0 && j < s.count && s[i] == s[j] {
            i -= 1
            j += 1
        }
        return j - i - 1
    }
}